pkg/builtins/poseidon.go

```diff
@@ -0,0 +1,80 @@
+package builtins
+
+import (
+	"github.com/lambdaclass/cairo-vm.go/pkg/lambdaworks"
+	starknet_crypto "github.com/lambdaclass/cairo-vm.go/pkg/starknet_crypto"
+	"github.com/lambdaclass/cairo-vm.go/pkg/vm/memory"
+)
+
+const POSEIDON_BUILTIN_NAME = "poseidon"
+const POSEIDON_CELLS_PER_INSTANCE = 6
+const POSEIDON_INPUT_CELLS_PER_INSTANCE = 3
+
+type PoseidonBuiltinRunner struct {
+	base     memory.Relocatable
+	included bool
+	cache    map[memory.Relocatable]lambdaworks.Felt
+}
+
+func NewPoseidonBuiltinRunner(included bool) *PoseidonBuiltinRunner {
+	return &PoseidonBuiltinRunner{included: included, cache: make(map[memory.Relocatable]lambdaworks.Felt)}
+}
+
+func (p *PoseidonBuiltinRunner) Base() memory.Relocatable {
+	return p.base
+}
+
+func (p *PoseidonBuiltinRunner) Name() string {
+	return POSEIDON_BUILTIN_NAME
+}
+
+func (p *PoseidonBuiltinRunner) InitializeSegments(segments *memory.MemorySegmentManager) {
+	p.base = segments.AddSegment()
+}
+
+func (p *PoseidonBuiltinRunner) InitialStack() []memory.MaybeRelocatable {
+	if p.included {
+		return []memory.MaybeRelocatable{*memory.NewMaybeRelocatableRelocatable(p.base)}
+	} else {
+		return nil
+	}
+}
+
+func (p *PoseidonBuiltinRunner) DeduceMemoryCell(address memory.Relocatable, mem *memory.Memory) (*memory.MaybeRelocatable, error) {
+	// Check if its an input cell
+	index := address.Offset % POSEIDON_CELLS_PER_INSTANCE
+	if index < POSEIDON_INPUT_CELLS_PER_INSTANCE {
+		return nil, nil
+	}
+
+	value, ok := p.cache[address]
+	if ok {
+		return memory.NewMaybeRelocatableFelt(value), nil
+	}
+
+	input_start_addr, _ := address.SubUint(index)
```

Comment:
I would add a comment on why we are discarding the error on the subtraction (I understand that's because `address <= index`, right?)

Reply:
sure